runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/data/BlockManagerWorker.java

```diff
@@ -166,28 +168,34 @@ public void createBlock(final String blockId,
       final DataStoreProperty.Value blockStore,
       final KeyRange keyRange) {
     // Let's see if a remote worker has it
-    // Ask Master for the location
-    final CompletableFuture<ControlMessage.Message> responseFromMasterFuture = persistentConnectionToMasterMap
-        .getMessageSender(MessageEnvironment.BLOCK_MANAGER_MASTER_MESSAGE_LISTENER_ID).request(
-            ControlMessage.Message.newBuilder()
-                .setId(RuntimeIdGenerator.generateMessageId())
-                .setListenerId(MessageEnvironment.BLOCK_MANAGER_MASTER_MESSAGE_LISTENER_ID)
-                .setType(ControlMessage.MessageType.RequestBlockLocation)
-                .setRequestBlockLocationMsg(
-                    ControlMessage.RequestBlockLocationMsg.newBuilder()
-                        .setExecutorId(executorId)
-                        .setBlockId(blockId)
-                        .build())
-                .build());
+    final CompletableFuture<ControlMessage.Message> blockLocationFuture =
+        pendingBlockLocationRequest.computeIfAbsent(blockId, blockIdToRequest -> {
+          // Ask Master for the location
+          final CompletableFuture<ControlMessage.Message> responseFromMasterFuture = persistentConnectionToMasterMap
+              .getMessageSender(MessageEnvironment.BLOCK_MANAGER_MASTER_MESSAGE_LISTENER_ID).request(
+                  ControlMessage.Message.newBuilder()
+                      .setId(RuntimeIdGenerator.generateMessageId())
+                      .setListenerId(MessageEnvironment.BLOCK_MANAGER_MASTER_MESSAGE_LISTENER_ID)
+                      .setType(ControlMessage.MessageType.RequestBlockLocation)
+                      .setRequestBlockLocationMsg(
+                          ControlMessage.RequestBlockLocationMsg.newBuilder()
+                              .setExecutorId(executorId)
+                              .setBlockId(blockId)
+                              .build())
+                      .build());
+          return responseFromMasterFuture;
+        });
+    blockLocationFuture.whenComplete((message, throwable) -> pendingBlockLocationRequest.remove(blockId));
```

Comment:
In my thought, if we want to not remove the future immediately, the master have to notice the state change of committed block to all executors to evict the location from their cache.

The major benefit of this caching comes when we doing shuffle push.
In this case, all of the receiver task groups are scheduled and request block location before the producer schedule groups are scheduled.
If we apply this caching, all receiver task groups will share the location future.

Reply:
Understood. More requests will overlap more in the case of a push-based shuffle. I guess then it'd be good to remove the future immediately to make things simple.

That said, note that the current approach also has the cache invalidation problem you pointed out, although the chance of it happening is smaller.

Reply:
You are right. If a block lost right after when it is committed, the receiver task groups will receive the stale location. After this, the fetch operation will fail and the task group will be retired through our fault tolerance mechanism.
However, this chance is not introduced by this caching, but exist already. Also, in my thought, it is very hard to avoid this chance unless we expect the machine failure.

Reply:
Right. I'll look into this when I run Pado experiments. 😄